## Profitable App Profiles for the App Store and Google Play Markets
My aim is to help developers understand what kinds of apps are likely to attract more users on Google Play and the App Store. To do this, I need to collect and analyze data about mobile apps available on Google Play and the App Store.

## Opening and Exploring the Data
As of September 2018, there were approximately 2 million iOS apps available on the App Store, and 2.1 million Android apps on Google Play.

Collecting data for over four million apps requires a significant amount of time and money, so I'll analyze a sample of the data. Luckily, these are two data sets that seem suitable for my goals:

* [A data set](https://www.kaggle.com/lava18/google-play-store-apps/home) containing data about approximately ten thousand Android apps from Google Play

* [A data set](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home) containing data about approximately seven thousand iOS apps from the App Store

I'll first open and explore tha data.

In [143]:
import pandas as pd
import numpy as np

In [144]:
#reading in the ios dataset
ios = pd.read_csv(r'C:\Users\Geoff\iCloudDrive\Documents\Data Science\Projects\APP Store Project\AppleStore.csv')

In [145]:
#reading in the android dataset
android = pd.read_csv(r'C:\Users\Geoff\iCloudDrive\Documents\Data Science\Projects\APP Store Project\googleplaystore.csv')

## iOS App Store dataset

In [146]:
#Looking at the ios column names
ios.columns

Index(['Unnamed: 0', 'id', 'track_name', 'size_bytes', 'currency', 'price',
       'rating_count_tot', 'rating_count_ver', 'user_rating',
       'user_rating_ver', 'ver', 'cont_rating', 'prime_genre',
       'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'],
      dtype='object')

In [147]:
ios.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7197 entries, 0 to 7196
Data columns (total 17 columns):
Unnamed: 0          7197 non-null int64
id                  7197 non-null int64
track_name          7197 non-null object
size_bytes          7197 non-null int64
currency            7197 non-null object
price               7197 non-null float64
rating_count_tot    7197 non-null int64
rating_count_ver    7197 non-null int64
user_rating         7197 non-null float64
user_rating_ver     7197 non-null float64
ver                 7197 non-null object
cont_rating         7197 non-null object
prime_genre         7197 non-null object
sup_devices.num     7197 non-null int64
ipadSc_urls.num     7197 non-null int64
lang.num            7197 non-null int64
vpp_lic             7197 non-null int64
dtypes: float64(3), int64(9), object(5)
memory usage: 955.9+ KB


In [148]:
ios.head()

Unnamed: 0         id                                         track_name  \
0           1  281656475                                    PAC-MAN Premium   
1           2  281796108                          Evernote - stay organized   
2           3  281940292    WeatherBug - Local Weather, Radar, Maps, Alerts   
3           4  282614216  eBay: Best App to Buy, Sell, Save! Online Shop...   
4           5  282935706                                              Bible   

   size_bytes currency  price  rating_count_tot  rating_count_ver  \
0   100788224      USD   3.99             21292                26   
1   158578688      USD   0.00            161065                26   
2   100524032      USD   0.00            188583              2822   
3   128512000      USD   0.00            262241               649   
4    92774400      USD   0.00            985920              5320   

   user_rating  user_rating_ver     ver cont_rating   prime_genre  \
0          4.0              4.5   6.3.5          4+         Games   
1          4.0              3.5   8.2.2          4+  Productivity   
2          3.5              4.5   5.0.0          4+       Weather   
3          4.0              4.5  5.10.0         12+      Shopping   
4          4.5              5.0   7.5.1          4+     Reference   

   sup_devices.num  ipadSc_urls.num  lang.num  vpp_lic  
0               38                5        10        1  
1               37                5        23        1  
2               37                5         3        1  
3               37                5         9        1  
4               37                5        45        1

## Google Play dataset

In [149]:
# looking at the columns
android.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [150]:
# Looking at the data types
android.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
App               10841 non-null object
Category          10841 non-null object
Rating            9367 non-null float64
Reviews           10841 non-null object
Size              10841 non-null object
Installs          10841 non-null object
Type              10840 non-null object
Price             10841 non-null object
Content Rating    10840 non-null object
Genres            10841 non-null object
Last Updated      10841 non-null object
Current Ver       10833 non-null object
Android Ver       10838 non-null object
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [151]:
android.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

### Observations: 
We see that the Google Play data set has 10841 apps and 13 columns. 

Usefull columns from the Google Play dataset are:

* App
* Category
* Genres
* Rating
* Reviews
* Installs
* Type
* Price

The iOS app store dataset has 7,197 apps and 17 columns.

Useful columns from the App Store dataset are:

* track_name
* rating_count_ver
* prime_genre
* rating_count_tot
* user_rating
* price


## Cleaning and Preparing the Data

### Android Play Dataset

There is a [discussion forum](https://www.kaggle.com/lava18/google-play-store-apps/discussion/81460) on Kaggle for this Dataset. It indicates there is an error in the Android dataset in row 10472. It also indicates there are duplicate rows.

### Deleting Wrong Data 
The [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) on Kaggle regarding this dataset that indicates there is an error for row 10472. I'll explore this row and evaluate whether or not it needs to be deleted.

In [152]:
android.iloc[10472]

App               Life Made WI-Fi Touchscreen Photo Frame
Category                                              1.9
Rating                                                 19
Reviews                                              3.0M
Size                                               1,000+
Installs                                             Free
Type                                                    0
Price                                            Everyone
Content Rating                                        NaN
Genres                                  February 11, 2018
Last Updated                                       1.0.19
Current Ver                                    4.0 and up
Android Ver                                           NaN
Name: 10472, dtype: object

In [153]:
# It seems the data has been shifted and is incorrect. It will be dropped from the dataset
android.drop(index=10472)

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
5                             Paper flowers instructions       ART_AND_DESIGN   
6                Smoke Effect Photo Maker - Smoke Editor       ART_AND_DESIGN   
7                                       Infinite Painter       ART_AND_DESIGN   
8                                   Garden Coloring Book       ART_AND_DESIGN   
9                          Kids Paint Free - Drawing Fun       ART_AND_DESIGN   
10                               Text on Photo - Fonteee       ART_AND_DESIGN   
11               Name Art Photo Editor - Focus n Filters       ART_AND_DESIGN   
12                        Tattoo Name On My Photo Editor       ART_AND_DESIGN   
13                                 Mandala Coloring Book       ART_AND_DESIGN   
14       3D Color Pixel by Number - Sandbox Art Coloring       ART_AND_DESIGN   
15                       Learn To Draw Kawaii Characters       ART_AND_DESIGN   
16          Photo Designer - Write your name with shapes       ART_AND_DESIGN   
17                              350 Diy Room Decor Ideas       ART_AND_DESIGN   
18                         FlipaClip - Cartoon animation       ART_AND_DESIGN   
19                                          ibis Paint X       ART_AND_DESIGN   
20                           Logo Maker - Small Business       ART_AND_DESIGN   
21             Boys Photo Editor - Six Pack & Men's Suit       ART_AND_DESIGN   
22               Superheroes Wallpapers | 4K Backgrounds       ART_AND_DESIGN   
23                                Mcqueen Coloring pages       ART_AND_DESIGN   
24                           HD Mickey Minnie Wallpapers       ART_AND_DESIGN   
25                            Harley Quinn wallpapers HD       ART_AND_DESIGN   
26                         Colorfit - Drawing & Coloring       ART_AND_DESIGN   
27                                 Animated Photo Editor       ART_AND_DESIGN   
28                                 Pencil Sketch Drawing       ART_AND_DESIGN   
29                       Easy Realistic Drawing Tutorial       ART_AND_DESIGN   
...                                                  ...                  ...   
10811                                        FR Plus 1.6    AUTO_AND_VEHICLES   
10812                                      Fr Agnel Pune               FAMILY   
10813                                     DICT.fr Mobile             BUSINESS   
10814                               FR: My Secret Pets!                FAMILY   
10815                          Golden Dictionary (FR-AR)  BOOKS_AND_REFERENCE   
10816                                 FieldBi FR Offline             BUSINESS   
10817                               HTC Sense Input - FR                TOOLS   
10818                               Gold Quote - Gold.fr              FINANCE   
10819                                          Fanfic-FR  BOOKS_AND_REFERENCE   
10820                                    Fr. Daoud Lamei               FAMILY   
10821                                            Poop FR               FAMILY   
10822                                          PLMGSS FR         PRODUCTIVITY   
10823                                       List iptv FR        VIDEO_PLAYERS   
10824                                          Cardio-FR              MEDICAL   
10825                                 Naruto & Boruto FR               SOCIAL   
10826          Frim: get new friends on local chat rooms               SOCIAL   
10827                                 Fr Agnel Ambarnath               FAMILY   
10828                            Manga-FR - Anime Vostfr               COMICS   
108

### Removing duplicates from Android

Due to the apps having multiple versions, there are many rows with duplicate names. I don’t want to count certain apps more than once when I analyze the data, so I need to remove the duplicate entries and keep only one entry per app. 

I'll keep the rows with the highest number of reviews, as I'm assuming that the higher the number of reviews, the more reliable the ratings.

In [154]:
android.shape

(10841, 13)

In [155]:
android.duplicated(subset="App", keep=False).value_counts()

False    8862
True     1979
dtype: int64

In [156]:
android[android.duplicated(subset='App')]

App            Category  \
229                         Quick PDF Scanner + OCR FREE            BUSINESS   
236                                                  Box            BUSINESS   
239                                   Google My Business            BUSINESS   
256                                  ZOOM Cloud Meetings            BUSINESS   
261                            join.me - Simple Meetings            BUSINESS   
265                                                  Box            BUSINESS   
266                                             Zenefits            BUSINESS   
267                                           Google Ads            BUSINESS   
268                                   Google My Business            BUSINESS   
269                                                Slack            BUSINESS   
270                                   FreshBooks Classic            BUSINESS   
271                                        Insightly CRM            BUSINESS   
272          QuickBooks Accounting: Invoicing & Expenses            BUSINESS   
273                       HipChat - Chat Built for Teams            BUSINESS   
274                             Xero Accounting Software            BUSINESS   
275              MailChimp - Email, Marketing Automation            BUSINESS   
276                 Crew - Free Messaging and Scheduling            BUSINESS   
277                        Asana: organize team projects            BUSINESS   
278                                     Google Analytics            BUSINESS   
279                                      AdWords Express            BUSINESS   
280                          Accounting App - Zoho Books            BUSINESS   
281                       Invoice & Time Tracking - Zoho            BUSINESS   
282                            join.me - Simple Meetings            BUSINESS   
283    Invoice 2go — Professional Invoices and Estimates            BUSINESS   
284     SignEasy | Sign and Fill PDF and other Documents            BUSINESS   
285                         Quick PDF Scanner + OCR FREE            BUSINESS   
286                            Genius Scan - PDF Scanner            BUSINESS   
287                       Tiny Scanner - PDF Scanner App            BUSINESS   
288                         Fast Scanner : Free PDF Scan            BUSINESS   
289                     Mobile Doc Scanner (MDScan) Lite            BUSINESS   
...                                                  ...                 ...   
9854       Nike Training Club - Workouts & Fitness Plans  HEALTH_AND_FITNESS   
9856                       Hangouts Dialer - Call Phones       COMMUNICATION   
9859                           Offline Maps & Navigation    TRAVEL_AND_LOCAL   
9936               Strawberry Shortcake Ice Cream Island              FAMILY   
9975                         Home Workout - No Equipment  HEALTH_AND_FITNESS   
9982   Home Security Camera WardenCam - reuse old phones      HOUSE_AND_HOME   
10018                                       Food Network              FAMILY   
10026                            Web Browser for Android       COMMUNICATION   
10049          Airway Ex - Intubate. Anesthetize. Train.             MEDICAL   
10118                      FilterGrid - Cam&Photo Editor         PHOTOGRAPHY   
10170        Messages, Text and Video Chat for Messenger              SOCIAL   
10171                                All Social Networks              SOCIAL   
10178                      Premier League - Official App              SPORTS   
10186                                   Farm Heroes Saga              FAMILY   
10188                                ESPN Fantasy Sports              SPORTS   
10190                                    Fallout Shelter              FAMILY   
10200                             Facebook Pages Manager            BUSINESS   
10203                               Facebook Ads Manager            BUSINESS   
10213  Who Viewed My Facebook Profile - Stalkers Visi

In [157]:
# Exploring duplicates. What are the differences? I picked one that had a high amout of downloands and high version number to evaluate.
android[android.App == 'Farm Heroes Saga']

App Category  Rating  Reviews Size      Installs  Type  \
1695   Farm Heroes Saga     GAME     4.4  7614130  70M  100,000,000+  Free   
1733   Farm Heroes Saga     GAME     4.4  7614271  70M  100,000,000+  Free   
1880   Farm Heroes Saga     GAME     4.4  7614415  70M  100,000,000+  Free   
2011   Farm Heroes Saga     GAME     4.4  7614407  70M  100,000,000+  Free   
10186  Farm Heroes Saga   FAMILY     4.4  7615646  71M  100,000,000+  Free   

      Price Content Rating  Genres    Last Updated Current Ver Android Ver  
1695      0       Everyone  Casual   July 26, 2018       5.1.8  2.3 and up  
1733      0       Everyone  Casual   July 26, 2018       5.1.8  2.3 and up  
1880      0       Everyone  Casual   July 26, 2018       5.1.8  2.3 and up  
2011      0       Everyone  Casual   July 26, 2018       5.1.8  2.3 and up  
10186     0       Everyone  Casual  August 7, 2018       5.2.6  2.3 and up

#### Observation:
It seems that the only difference is the number of reviews. This is likely due to the data being collected at different times. I'll keep only the entries with the largest number of reviews.

In [158]:
# Using pandas to sort the values of the duplicates by number of reviews and only keeping only the highest.

android = android.sort_values('Reviews', ascending=True).drop_duplicates('App', keep='last').sort_index()

In [159]:
# Checking for accuracy
android[android.App == 'Farm Heroes Saga']

App Category  Rating  Reviews Size      Installs  Type  \
10186  Farm Heroes Saga   FAMILY     4.4  7615646  71M  100,000,000+  Free   

      Price Content Rating  Genres    Last Updated Current Ver Android Ver  
10186     0       Everyone  Casual  August 7, 2018       5.2.6  2.3 and up

In [160]:
android.shape

(9660, 13)

### Removing non English Apps
Some of the app names suggest they are not directed toward an English-speaking audience. It is difficult to remove non english app names in a perfect manner. One popular method is to remove any entries with three or more non-standard (ordinal above 127) ASCII characters, though some apps with symbols or emoji could get caught up in this filtering method. While not perfect, it is better than other methods. 

In [161]:
import string

def nonEnglishCharacterCount(app_name):
    non_eng_char_ct = 0
    for character in app_name:
        if ord(character) > 127:
            non_eng_char_ct += 1
    return non_eng_char_ct

In [162]:
android['num_non_eng_chars'] = [nonEnglishCharacterCount(i) for i in android['App']]

In [163]:
android = android.loc[android['num_non_eng_chars'] <= 3]
android.shape

(9615, 14)

## iOS Data Set

#### removing iOS duplicates

In [164]:
ios.shape

(7197, 17)

In [165]:
ios.duplicated(subset='track_name', keep=False).value_counts()


False    7193
True        4
dtype: int64

In [166]:
ios[ios.duplicated(subset='track_name')]

Unnamed: 0          id           track_name  size_bytes currency  price  \
5603        7579  1089824278    VR Roller Coaster   240964608      USD    0.0   
7128       10885  1178454060  Mannequin Challenge    59572224      USD    0.0   

      rating_count_tot  rating_count_ver  user_rating  user_rating_ver    ver  \
5603                67                44          3.5              4.0   0.81   
7128               105                58          4.0              4.5  1.0.1   

     cont_rating prime_genre  sup_devices.num  ipadSc_urls.num  lang.num  \
5603          4+       Games               38                0         1   
7128          4+       Games               38                5         1   

      vpp_lic  
5603        1  
7128        1

In [167]:
ios = ios.sort_values('rating_count_tot', ascending=True).drop_duplicates('track_name', keep='last').sort_index()

In [168]:
ios.shape

(7195, 17)

### Removing non English Apps

In [169]:
ios['num_non_eng_chars'] = [nonEnglishCharacterCount(i) for i in ios['track_name']]
ios = ios[ios['num_non_eng_chars'] <= 3]
ios.shape

(6181, 18)

## Output datasets to excel for further analysis in Tableau

In [170]:
ios.to_csv('ios.csv')
ios.to_excel('ios.xlsx')
android.to_csv('android.csv')
android.to_excel('android.xlsx')